lib/scholar/cluster/affinity_propagation.ex

```diff
@@ -0,0 +1,378 @@
+  Computes labels and centers of clusterization.
+
+  ## Return Values
+
+    The function returns a struct with the following parameters:
+
+    * `:affinity_matrix` - Affinity matrix. It is a negated squared euclidean distance of each pair of points.
+
+    * `:clusters_centers` - Cluster centers from the initial data.
+
+    * `:cluster_centers_indices` - Indices of cluster centers.
+
+    * `:labels` - Labels of each point.
+
+    * `:num_clusters` - Number of clusters.
+
+    * `:similarity_matrix` - Similarity matrix. It is a similarity of each pair of points.
+
+  ## Examples
+
+      iex> seed = 42
+      iex> x = Nx.tensor([[12,5,78,2], [1,-5,7,32], [-1,3,6,1], [1,-2,5,2]])
+      iex> model = Scholar.Cluster.AffinityPropagation.fit(x, seed: seed)
+      iex> Scholar.Cluster.AffinityPropagation.prune(model)
+      %Scholar.Cluster.AffinityPropagation{
+        labels: Nx.tensor([0, 1, 1, 1]),
+        cluster_centers_indices: Nx.tensor([0, 3]),
+        affinity_matrix: Nx.tensor(
+          [
+            [0.0, -6162.0, -5358.0, -5499.0],
+            [-6162.0, 0.0, -1030.0, -913.0],
+            [-5358.0, -1030.0, 0.0, -31.0],
+            [-5499.0, -913.0, -31.0, 0.0]
+          ]),
+        cluster_centers: Nx.tensor(
+          [
+            [12, 5, 78, 2],
+            [1, -2, 5, 2]
+          ]
+        ),
+        num_clusters: Nx.tensor(2, type: :u64),
+        similarity_matrix: Nx.tensor(
+          [
+            [-971.5, -6162.0, -5358.0, -5499.0],
+            [-6162.0, -971.5, -1030.0, -913.0],
+            [-5358.0, -1030.0, -971.5, -31.0],
+            [-5499.0, -913.0, -31.0, -971.5]
+          ]
+        )
+      }
+  """
+  def prune(
+        %__MODULE__{
+          cluster_centers_indices: cluster_centers_indices,
+          cluster_centers: cluster_centers,
+          num_clusters: k,
+          similarity_matrix: similarity_matrix
+        } = model
+      ) do
+    k = Nx.to_number(k)
+
+    cluster_centers_indices =
+      Nx.to_flat_list(cluster_centers_indices) |> Enum.filter(&(&1 != -1)) |> Nx.tensor()
+
+    cluster_centers = Nx.take(cluster_centers, cluster_centers_indices)
+
+    labels = Nx.argmax(Nx.take(similarity_matrix, cluster_centers_indices, axis: 1), axis: 1)
+    labels = Nx.indexed_put(labels, Nx.new_axis(cluster_centers_indices, -1), Nx.iota({k}))
+
+    %__MODULE__{
+      model
+      | cluster_centers_indices: cluster_centers_indices,
+        cluster_centers: cluster_centers,
+        labels: labels
+    }
+  end
+
+  @doc """
+  Predict the closest cluster each sample in `x` belongs to.
+
+  ## Examples
+
+      iex> seed = 42
+      iex> x = Nx.tensor([[12,5,78,2], [1,5,7,32], [1,3,6,1], [1,2,5,2]])
+      iex> model = Scholar.Cluster.AffinityPropagation.fit(x, seed: seed)
+      iex> model = Scholar.Cluster.AffinityPropagation.prune(model)
+      iex> Scholar.Cluster.AffinityPropagation.predict(model, Nx.tensor([[1,6,2,6], [8,3,8,2]]))
+      #Nx.Tensor<
+        s64[2]
+        [1, 1]
```

Comment:
So what is the meaning of this result? The 1 refers to positions in `x`? Can we clarify it?

Reply:
This means that these two samples were labelled to the first centroid (at index 1 in clusters_centers after calling function prune)